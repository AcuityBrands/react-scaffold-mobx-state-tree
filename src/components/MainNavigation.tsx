/**
 * Primary Navigation Menu
 * 
 * This component will listen for changes to the router location
 * and highlight the appropriate menu keys appropriately
 * 
 * Warning: Dependent on dom router
 */

import * as React from "react";
import { withRouter } from 'react-router-dom';
import { Link } from 'react-router-dom';
import { Menu, Icon } from 'antd';
import MenuItem from "antd/lib/menu/MenuItem";

interface IMainNavProps {
  match: any,
  location: any,
  history: any
}

interface IMenuItem {
  key: string,
  path: string,
  icon: string,
  label: string
}

class MainNavigation extends React.Component<IMainNavProps, undefined>{
  items: Array<IMenuItem> = [
    { key: "1", path: "/dashboard", icon: "line-chart", label: "Dashboard" },
    { key: "2", path: "/task", icon: "check-circle-o", label: "Tasks" },
    { key: "3", path: "/fail", icon: "star-o", label: "Sample 404" }
  ]
  selectedKeys:Array<string> = [];
  unlisten:any;

  componentDidMount() {
    this.selectKey(this.props.location);
    this.unlisten = this.props.history.listen((location:any, action:any)=>{
      this.selectKey(location);
    });
  }

  selectKey = (location:any) => {
    this.selectedKeys = [];
    this.items.map(i => {
      if(location.pathname == i.path){
        this.selectedKeys.push(i.key);
      }
    })
  }

  componentWillUnmount() {
      this.unlisten();
  }

  render() {
    // Create the menu with menu items
    return (
      <Menu mode="inline"
        theme="dark"
        selectedKeys={this.selectedKeys}
      >
        {this.items.map((i) => {
          return <Menu.Item key={i.key}>
            <Link to={i.path}><Icon type={i.icon} /><span>{i.label}</span></Link>
          </Menu.Item>
        })}
        <Menu.Divider className="nav-divider"/>
        <Menu.Item key="99">
          <Link to="/about"><Icon type="info-circle-o" /><span>About Application</span></Link>
        </Menu.Item>
      </Menu>
      );
    }
  }
  
  //TODO: This has a dependency on the react-router
  export default withRouter(MainNavigation);