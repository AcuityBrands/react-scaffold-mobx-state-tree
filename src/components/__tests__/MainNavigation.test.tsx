import * as React from "react";
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { MainNavigation } from '../MainNavigation';

Enzyme.configure({ adapter: new Adapter() });

const setup = propOverrides => {
  const props = Object.assign({
    match: {},
    location: {},
    history:{
      listen: jest.fn()
    },
    config: [
      { key: "1", path: "/dashboard", icon: "test", label: "test", component: "test"},
      { key: "2", path: "/xxx", icon: "test", label: "test", component: "test"}
    ]
  }, propOverrides)
  const wrapper = Enzyme.shallow(<MainNavigation {...props} />)
  return {
    props,
    wrapper
  }
}

describe('Main Navigation Component', () => {
  // Minimal component test confirms component rendered
  it("can render", () => {
    const { wrapper, props} = setup({});
    expect(wrapper.exists()).toBe(true);
  })

  it("matches the snapshot", () => {
    const { wrapper } = setup({});
    expect(wrapper).toMatchSnapshot();
  })

  it("always renders a Menu", () => {
    const { wrapper } = setup({});
    const menu = wrapper.find("Menu");
    expect(menu.length).toBeGreaterThan(0);
  });

  it("sets the selected key based off the location", () => {
    const { wrapper } = setup({location:{pathname:"/dashboard"}});
    expect(wrapper.instance().selectedKeys.length).toBe(1);
  });

  it("does not select a key if no key is found", () => {
    const { wrapper } = setup({location:{pathname:"/garbage"}});
    expect(wrapper.instance().selectedKeys.length).toBe(0);
  });

  it("should create a menu item for each item in config", () => {
    const { wrapper } = setup({});
    const items = wrapper.find("Menu").children();
    expect(items.length).toBe(2);
  });

  it("should create a divider if component == 'Divider'", () => {
    const { wrapper } = setup({config:[
      { key: "2", path: "/xxx", icon: "test", label: "test", component: "Divider"}
    ]});
    expect(wrapper.find("Menu").children().length).toBe(1);
    expect(wrapper.find(".nav-divider").length).toBe(1);
  });
})