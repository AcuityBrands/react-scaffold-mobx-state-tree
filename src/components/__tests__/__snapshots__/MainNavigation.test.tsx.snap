// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Navigation Component matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainNavigation
    config={
      Array [
        Object {
          "component": "test",
          "icon": "test",
          "key": "1",
          "label": "test",
          "path": "/dashboard",
        },
        Object {
          "component": "test",
          "icon": "test",
          "key": "2",
          "label": "test",
          "path": "/xxx",
        },
      ]
    }
    history={
      Object {
        "listen": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
        },
      }
    }
    location={Object {}}
    match={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <MenuItem>
          <Link
            replace={false}
            to="/dashboard"
          >
            <Icon
              type="test"
            />
            <span>
              test
            </span>
          </Link>
        </MenuItem>,
        <MenuItem>
          <Link
            replace={false}
            to="/xxx"
          >
            <Icon
              type="test"
            />
            <span>
              test
            </span>
          </Link>
        </MenuItem>,
      ],
      "className": "",
      "mode": "inline",
      "prefixCls": "ant-menu",
      "selectedKeys": Array [],
      "theme": "dark",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "children": <Link
            replace={false}
            to="/dashboard"
          >
            <Icon
              type="test"
            />
            <span>
              test
            </span>
          </Link>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Icon
                type="test"
              />,
              <span>
                test
              </span>,
            ],
            "replace": false,
            "to": "/dashboard",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "type": "test",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "test",
              },
              "ref": null,
              "rendered": "test",
              "type": "span",
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "children": <Link
            replace={false}
            to="/xxx"
          >
            <Icon
              type="test"
            />
            <span>
              test
            </span>
          </Link>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Icon
                type="test"
              />,
              <span>
                test
              </span>,
            ],
            "replace": false,
            "to": "/xxx",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "type": "test",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "test",
              },
              "ref": null,
              "rendered": "test",
              "type": "span",
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MenuItem>
            <Link
              replace={false}
              to="/dashboard"
            >
              <Icon
                type="test"
              />
              <span>
                test
              </span>
            </Link>
          </MenuItem>,
          <MenuItem>
            <Link
              replace={false}
              to="/xxx"
            >
              <Icon
                type="test"
              />
              <span>
                test
              </span>
            </Link>
          </MenuItem>,
        ],
        "className": "",
        "mode": "inline",
        "prefixCls": "ant-menu",
        "selectedKeys": Array [],
        "theme": "dark",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "children": <Link
              replace={false}
              to="/dashboard"
            >
              <Icon
                type="test"
              />
              <span>
                test
              </span>
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Icon
                  type="test"
                />,
                <span>
                  test
                </span>,
              ],
              "replace": false,
              "to": "/dashboard",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "type": "test",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "test",
                },
                "ref": null,
                "rendered": "test",
                "type": "span",
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "children": <Link
              replace={false}
              to="/xxx"
            >
              <Icon
                type="test"
              />
              <span>
                test
              </span>
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Icon
                  type="test"
                />,
                <span>
                  test
                </span>,
              ],
              "replace": false,
              "to": "/xxx",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "type": "test",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "test",
                },
                "ref": null,
                "rendered": "test",
                "type": "span",
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
